### PR TITLE
fix: preserve directory hierarchy for installed headers

### PR DESCRIPTION
**Pull request by rgrinberg**
If we have foo/bar.h, we'll preserve the hierarchy relative to the
library directory

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

### CODE REVIEW

File: src/dune_rules/install_rules.ml [hunk 1]
```diff
@@ -279,8 +279,21 @@ end = struct
...
+        let path base =
+          Path.Build.relative dir (base ^ Foreign_language.header_extension)
+        in
+        let project = Scope.project scope in
+        if Dune_project.dune_version project < (3, 8) then fun base ->
+          make_entry Lib (path base)
+        else fun base ->
+          let sub_dir =
+            let dirname = Filename.dirname base in
+            if dirname = "." then None else in_sub_dir (Some dirname)
```
Comment (anmonteiro) [inline src/dune_rules/install_rules.ml:292]:
```suggestion
            if dirname = Filename.current_dir_name then None else in_sub_dir (Some dirname)
```

**Review by anmonteiro — APPROVED** [review]

**Comment by rgrinberg** [comment]
Let's have a look at @emillon as well

**Comment by rgrinberg** [comment]
On second thought, this fix is a little bit weird. For example, how would we handle something like `(install_c_headers ../foo)`?

Probably simpler to introduce the syntax from the `install` stanza and allow users to select the installed path. E.g. `(install_c_headers (./foo as baz))`. 

To make the common case simple, we could introduce a new constructor such as `(install_c_headers (with_subdir foo/bar))` that will have the same behavior as this PR.

**Comment by rgrinberg** [comment]
Okay, I just switched to using normal `(foo as bar)` patterns.

It requires manual work to get the install layout we want, but at least it's obvious and consistent with the rest of dune.

**Comment by rgrinberg** [comment]
Actually, I'm now leaning towards introducing a new field now. One thing I dislike about this implementation is that it makes globbing awkard. E.g. one cannot do `(install_c_headers (glob_files *.h))` for seemingly no good reason.

The new field is just going to be `public_headers` and it will allow proper filenames. As a bonus, this will allow header extensions such as `.hh` and `.hpp` that should be usable as well.

**Comment by rgrinberg** [comment]
@anmonteiro take another look please

**Comment by rgrinberg** [comment]
@anmonteiro I rewrote this feature again. This time with the realization that this almost exactly the same as melange's runtime deps feature.

File: src/dune_rules/install_rules.ml [hunk 2]
```diff
@@ -193,6 +192,8 @@ end = struct
           in
           make_entry ?sub_dir Lib path)
     in
+    let* melange_runtime_entries = additional_deps lib.melange_runtime_deps in
+    let* public_headers = additional_deps lib.public_headers in
```
Comment (anmonteiro) [inline src/dune_rules/install_rules.ml:196]: perhaps this can be done in parallel? `let* ... and* ... in`

**Review by anmonteiro — APPROVED** [review]
Looks great. It looks very similar to runtime_deps indede.

File: src/dune_rules/install_rules.ml (hunk 2 above)
Comment (rgrinberg) [reply src/dune_rules/install_rules.ml:196]: Done. There's many opportunities to do such optimizations in the install rules.

Comment (Alizter) [reply src/dune_rules/install_rules.ml]: Is this really an optimisation? I thought and* is just implemented as a bind.

File: src/dune_rules/lib_flags.ml [hunk 3]
```diff
@@ -152,7 +152,35 @@ module L = struct
...
+              (* TODO handle sandboxing *)
+              deps
+            in
+            (headers :: local, external_))
+    in
+    let local =
+      let open Action_builder.O in
+      let* bindings = Action_builder.of_memo @@ Memo.all_concurrently local in
+      let+ () = Action_builder.all_unit bindings in
+      Command.Args.S []
```
Comment (emillon) [inline src/dune_rules/lib_flags.ml:180]:
```suggestion
      Command.Args.empty
```

File: src/dune_rules/dune_file.ml [hunk 4]
```diff
@@ -599,6 +600,11 @@ module Library = struct
        and+ synopsis = field_o "synopsis" string
        and+ install_c_headers =
          field "install_c_headers" (repeat (located string)) ~default:[]
+       and+ public_headers =
+         field "public_headers"
```
Comment (emillon) [inline src/dune_rules/dune_file.ml:604]: Can you add some doc about this in `doc/stanzas/library.rst`?

**Review by emillon — APPROVED** [review]
Thanks. That's pretty much what I had in mind in an early implementation.

File: src/dune_rules/install_rules.ml (hunk 2 above)
Comment (rgrinberg) [reply src/dune_rules/install_rules.ml]: Look at the definition of `and*` in fiber.

**Comment by rgrinberg** [comment]
@anmonteiro i've now forbidden all non file deps in public_headers. A similar change to melange runtime deps could be done.

**Comment by anmonteiro** [comment]
Thanks for the heads up, I’ll apply the fix there too.

**Comment by aantron** [comment]
As I understand it, `public_headers` as implemented here is not sufficient for Luv.

I have a library in `src/c` that needs to install headers from `src/c/vendor/include/**` such that they are accessible inside a monorepo through `-I` flags passed by Dune to sibling packages that have C code that depends on the library in `src/c` without having to include the `vendor/include` path fragment inside `#include`. That is, I need either the headers to be copied to the install directory for the library in `src/c`, or `-I that-directory/vendor/include` to be passed to the C bindings in the sibling package. Doing `dune build` shows that these flags are not passed to the C code being compiled in the sibling library, and, predictably, that library cannot find the headers.

**Comment by aantron** [comment]
More information here: https://github.com/ocaml/dune/issues/4121#issuecomment-1849417965

**Pull request merged**